### PR TITLE
JDK-8256475: Fix Behavior when Installer name differs from application name.

### PR DESCRIPTION
**Pull request by andyherrick**
…n name.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (6/6 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8256475](https://bugs.openjdk.java.net/browse/JDK-8256475): Fix Behavior when Installer name differs from application name.

### Reviewers
 * [Alexey Semenyuk](https://openjdk.java.net/census#asemenyuk) (@alexeysemenyukoracle - Committer)
 * [Alexander Matveev](https://openjdk.java.net/census#almatvee) (@sashamatveev - Committer)
 * [Alexander Zuev](https://openjdk.java.net/census#kizune) (@azuev-java - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1300/head:pull/1300`
`$ git checkout pull/1300`

### CODE REVIEW

File: src/jdk.jpackage/windows/classes/jdk/jpackage/internal/WinMsiBundler.java [hunk 1]
```diff
@@ -305,23 +306,21 @@ public boolean validate(Map<String, ? super Object> params)
     private void prepareProto(Map<String, ? super Object> params)
                 throws PackagerException, IOException {
         Path appImage = StandardBundlerParam.getPredefinedAppImage(params);
+        String appName = APP_NAME.fetchFrom(params);
         Path appDir;
-        String appName;
+        if (appName == null) {
+            // this can only happen when no name is given and using forign app-image
```
Comment (sashamatveev) [inline src/jdk.jpackage/windows/classes/jdk/jpackage/internal/WinMsiBundler.java:312]: forign -> foreign

File: test/jdk/tools/jpackage/share/MultiNameTwoPhaseTest.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import java.io.IOException;
+import jdk.jpackage.test.PackageTest;
+import jdk.jpackage.test.PackageType;
+import jdk.jpackage.test.TKit;
+import jdk.jpackage.test.Annotations.Test;
+import jdk.jpackage.test.Annotations.Parameter;
+import jdk.jpackage.test.JPackageCommand;
+
+/**
+ */
```
Comment (sashamatveev) [inline test/jdk/tools/jpackage/share/MultiNameTwoPhaseTest.java:34]: Needs description for test.

**Review by alexeysemenyukoracle — APPROVED** [review]

**Review by sashamatveev — APPROVED** [review]

**Comment by andyherrick** [comment]
@azuev-java - can you review as well ?

File: src/jdk.jpackage/share/classes/jdk/jpackage/internal/resources/MainResources_zh_CN.properties [hunk 3]
```diff
@@ -72,6 +72,7 @@ error.tool-old-version=\u627E\u4E0D\u5230 {0} {1}\u6216\u66F4\u65B0\u7248\u672C
 error.tool-old-version.advice=\u8BF7\u5B89\u88C5 {0} {1}\u6216\u66F4\u65B0\u7248\u672C
 error.jlink.failed=jlink \u5931\u8D25\uFF0C\u51FA\u73B0 {0}
 error.blocked.option=\u4E0D\u5141\u8BB8\u5728 --jlink-options \u4E2D\u4F7F\u7528 jlink \u9009\u9879 [{0}]
+error.no.name=Name not specified with --name and cannot infer one from app-image
```
Comment (azuev-java) [inline src/jdk.jpackage/share/classes/jdk/jpackage/internal/resources/MainResources_zh_CN.properties:75]: Do we need to submit a follow-up bug for localization team that we have new non-localized messages in the region specific properties file? I see that foreign app image warning is still not localized.

**Review by azuev-java — APPROVED** [review]
Looks good.

File: src/jdk.jpackage/share/classes/jdk/jpackage/internal/resources/MainResources_zh_CN.properties (hunk 3 above)
Comment (andyherrick) [reply src/jdk.jpackage/share/classes/jdk/jpackage/internal/resources/MainResources_zh_CN.properties:75]: Localization team will translate all changed messages together without individual follow-up bugs, as long as changed before the "localization freeze"

**Pull request closed without merging**